brainiak/utils/fmrisim.py

```diff
@@ -2134,126 +2701,229 @@ def generate_noise(dimensions,
+    # If you have only one magnitude value, duplicate the magnitude for each
+    #  timecourse you have
+    if len(magnitude) == 1:
+        magnitude *= signal_function.shape[1]
 
-    # Threshold the data
-    threshold = np.percentile(brain.reshape(np.prod(brain.shape[0:3])),
-                              percentile)
+    # Scale all signals that to have a range of -1 to 1. This is
+    # so that any values less than this will be scaled appropriately
+    signal_function /= np.max(np.abs(signal_function))
 
-    # How many voxels exceed a threshold
-    brain_threshold = np.where(np.abs(brain) > threshold)
+    # Iterate through the timecourses and calculate the metric
+    signal_function_scaled = np.zeros(signal_function.shape)
+    for voxel_counter in range(signal_function.shape[1]):
 
-    # Clear the way
-    ax.clear()
+        # Pull out the values for this voxel
+        sig_voxel = signal_function[:, voxel_counter]
+        noise_voxel = noise_function[:, voxel_counter]
+        magnitude_voxel = magnitude[voxel_counter]
 
-    ax.set_xlim(0, brain.shape[0])
-    ax.set_ylim(0, brain.shape[1])
-    ax.set_zlim(0, brain.shape[2])
+        # Calculate the scaled time course using the specified method
+        if method == 'SFNR':
 
-    # If a mask is provided then plot this
-    if mask is not None:
-        mask_threshold = np.where(np.abs(mask) > 0)
-        ax.scatter(mask_threshold[0],
-                   mask_threshold[1],
-                   mask_threshold[2],
-                   zdir='z',
-                   c='black',
-                   s=10,
-                   alpha=0.01)
-
-    # Plot the volume
-    ax.scatter(brain_threshold[0],
-               brain_threshold[1],
-               brain_threshold[2],
-               zdir='z',
-               c='red',
-               s=20)
-
-    return ax
+            # How much temporal variation is there, relative to the mean
+            # activity
+            temporal_var = noise_voxel.mean() / noise_dict['sfnr']
+
+            # Multiply the timecourse by the variability metric
+            new_sig = sig_voxel * (temporal_var * magnitude_voxel)
+
+        elif method == 'CNR_Amp/Noise-SD':
+
+            # What is the standard deviation of the noise
+            noise_std = np.std(noise_voxel)
+
+            # Multiply the signal timecourse by the the CNR and noise (
+            # rearranging eq.)
+            new_sig = sig_voxel * (magnitude_voxel * noise_std)
+
+        elif method == 'CNR_Amp2/Noise-Var_dB':
+
+            # Calculate the current signal amplitude (likely to be 1,
+            # but not necessarily)
+            sig_amp = np.max(np.abs(sig_voxel))
+
+            # What is the standard deviation of the noise
+            noise_std = np.std(noise_voxel)
+
+            # Rearrange the equation to compute the size of signal change in
+            #  decibels
+            scale = 10 ** ((magnitude_voxel / sig_amp) + np.log10(noise_std
```

Comment:
I think `(magnitude_voxel / sig_amp)` and `np.log10(noise_std ** 2)` are not compatible to be both in the power. Only magnitude_voxel is defined on log scale (since you are using dB) but sig_map and noise_std should not be used as power for 10.
I think it should be `(10 ** (magnitude_voxel/20)) * noise_std`. And if you want to scale signal_function by the maximum absolute value separately for different voxels, then `(10 ** (magnitude_voxel/20)) * noise_std / sig_amp`. The division by 20 is because the name of the metric seems to indicate that the dB value is calculated based on power: dB=10log(Power1/Power2)=20log(magnitude1/magnitude2)
And if you use this formula, then the next line does not need to take sqrt.
But anyway I think it is worth double checking the formula with some example simulated time courses.

Reply:
Good point, changed according and checked with the simulation below:

import numpy as np
import matplotlib.pyplot as plt
from brainiak.utils.fmrisim import _double_gamma_hrf

def sim_deci(decibels):
    noise_size=10
    sig_voxel=np.asarray(_double_gamma_hrf())
    noise_voxel=np.random.randn(len(sig_voxel)) * noise_size
    noise_std = np.std(noise_voxel)

    scale = (10 ** (decibels / 20)) * noise_std

    func = (sig_voxel * scale) + noise_voxel

    plt.plot(func)

# Amplitude should be scaled appropriately
sim_deci(10)
sim_deci(20)
sim_deci(30)